### ISSUE TITLE
[JavaScript] If Serialized by JS, Deserialized in Java will failed 

### ISSUE THREAD

**Issue by 3dot141** [alipay/fury#703]
**Describe the bug**
rt, i can't use it in js and java.

**Expected behavior**
if fury support cross platform, the action will be success

**To Reproduce**
js code 
```js
import Fury, {InternalSerializerType,  Type} from '@furyjs/fury';
import * as fs from "fs";

function main1() {

  const description = Type.object("example.bar", {
    c: Type.array(Type.object("example.foo", {
      a: Type.string()
    }))
  });
  const fury = new Fury();
  const {serialize, deserialize} = fury.registerSerializer(description);
  const o = {a: "123"};
  const rawObject = {c: [o, o]}
  console.log(`rawObject: ${JSON.stringify(rawObject)}`);
  const data = serialize({c: [o, o]})

  // create data directory
  if (!fs.existsSync('data')) {
    fs.mkdirSync('data');
  }
  // write data to file 'test.data'
  fs.writeFileSync('data/test.data', data);

  const deserialized = deserialize(data)
  console.log(`deserialized: ${JSON.stringify(deserialized)}`)
}
```

java code
```java
import cn.hutool.core.io.FileUtil;
import io.fury.Fury;
import io.fury.Language;

/**
 * @author Harrison
 * @since 11.0
 * Created on 2023/7/19
 */
public class FuryTest {
    
    public static class Test {
        
        A[] c;
    }
    
    public static class A {
        
        public String a;
    }
    
    public static void main(String[] args) {
        
        Fury fury = Fury.builder().withLanguage(Language.XLANG)
                .withRefTracking(true)
                .withSecureMode(false)
                .build();
        fury.register(Test.class, "example.bar");
        fury.register(A.class, "example.foo");
        
        byte[] bytes = FileUtil.readBytes("/Users/3dot141/Work/w_source/MagicKit/data/test.data");
        // bytes can be data serialized by other languages.
        
        Test test = new Test();

        A a1 = new A();
        a1.a = "123";
        A a2 = new A();
        a2.a = "123";

        test.c = new A[]{a1, a2};
        byte[] serialize = fury.serialize(test);
        
        Object deserialize = fury.deserialize(bytes);
    ;
        System.out.println();
    
    }
}

```

**Screenshots**
<img width="1165" alt="image" src="https://github.com/alipay/fury/assets/19152476/676d4c8d-bd0c-47fa-8e43-d4517b14e040">

**Environment (please complete the following information):**

**Comment by chaokunyang** [comment]
@wangweipeng2 Could you take a loot at javascript code?

**Comment by 3dot141** [comment]
i find one question

<img width="472" alt="image" src="https://github.com/alipay/fury/assets/19152476/5dceb25d-f729-4b02-8b57-49b49066be96">

js is "do while", but java is "while". 

when i fix it, i find array is error in js.

<img width="422" alt="image" src="https://github.com/alipay/fury/assets/19152476/0b138468-1536-4967-ab86-626b22539be5">

array type is 25 in js.

but in java, array type will be -25 .
<img width="384" alt="image" src="https://github.com/alipay/fury/assets/19152476/2048b317-f227-4f78-805e-b67eb49f7974">

**Comment by wangweipeng2** [comment]
> **Describe the bug** rt, i can't use it in js and java.
> 
> **Expected behavior** if fury support cross platform, the action will be success
> 
> **To Reproduce** js code
> 
> ```js
> import Fury, {InternalSerializerType,  Type} from '@furyjs/fury';
> import * as fs from "fs";
> 
> function main1() {
> 
>   const description = Type.object("example.bar", {
>     c: Type.array(Type.object("example.foo", {
>       a: Type.string()
>     }))
>   });
>   const fury = new Fury();
>   const {serialize, deserialize} = fury.registerSerializer(description);
>   const o = {a: "123"};
>   const rawObject = {c: [o, o]}
>   console.log(`rawObject: ${JSON.stringify(rawObject)}`);
>   const data = serialize({c: [o, o]})
> 
>   // create data directory
>   if (!fs.existsSync('data')) {
>     fs.mkdirSync('data');
>   }
>   // write data to file 'test.data'
>   fs.writeFileSync('data/test.data', data);
> 
>   const deserialized = deserialize(data)
>   console.log(`deserialized: ${JSON.stringify(deserialized)}`)
> }
> ```
> 
> java code
> 
> ```java
> import cn.hutool.core.io.FileUtil;
> import io.fury.Fury;
> import io.fury.Language;
> 
> /**
>  * @author Harrison
>  * @since 11.0
>  * Created on 2023/7/19
>  */
> public class FuryTest {
>     
>     public static class Test {
>         
>         A[] c;
>     }
>     
>     public static class A {
>         
>         public String a;
>     }
>     
>     public static void main(String[] args) {
>         
>         Fury fury = Fury.builder().withLanguage(Language.XLANG)
>                 .withRefTracking(true)
>                 .withSecureMode(false)
>                 .build();
>         fury.register(Test.class, "example.bar");
>         fury.register(A.class, "example.foo");
>         
>         byte[] bytes = FileUtil.readBytes("/Users/3dot141/Work/w_source/MagicKit/data/test.data");
>         // bytes can be data serialized by other languages.
>         
>         Test test = new Test();
> 
>         A a1 = new A();
>         a1.a = "123";
>         A a2 = new A();
>         a2.a = "123";
> 
>         test.c = new A[]{a1, a2};
>         byte[] serialize = fury.serialize(test);
>         
>         Object deserialize = fury.deserialize(bytes);
>     ;
>         System.out.println();
>     
>     }
> }
> ```
> 
> **Screenshots** <img alt="image" width="1165" src="https://user-images.githubusercontent.com/19152476/254539033-676d4c8d-bd0c-47fa-8e43-d4517b14e040.png">
> 
> **Environment (please complete the following information):**

Thank you for submitting this issue. We are still in the early stage, and the cross-language protocol may be adjusted at any time. Today, I will add cross-language protocol testing to ensure that there are no problems in the future.

**Comment by chaokunyang** [comment]
> i find one point
> 
> <img alt="image" width="472" src="https://user-images.githubusercontent.com/19152476/254568915-5dceb25d-f729-4b02-8b57-49b49066be96.png">
> js is "do while", but java is "while".
> 
> when i fix it, i find array is error in js.
> 
> <img alt="image" width="422" src="https://user-images.githubusercontent.com/19152476/254569399-0b138468-1536-4967-ab86-626b22539be5.png">
> array type is 25 in js.
> 
> but in java, array type will be -25 . <img alt="image" width="384" src="https://user-images.githubusercontent.com/19152476/254569619-2048b317-f227-4f78-805e-b67eb49f7974.png">

Good catch, I don't expect someone will dive into the code details. Xlang string serialization use utf8, but fury JavaScript supports latin and utf16 too. This may be the reason why deserialization failed. Fury Java supports those encoding，but We don't use Latin/uuf16 for Python， so we don't enable Latin/utf16 for java.

**Comment by chaokunyang** [comment]
We should add Latin/utf16 support for clang as soon as possible

**Comment by wangweipeng2** [comment]
```java
package io.fury;

import java.io.IOException;

public class FuryTest {

  public static class Test {

    A[] c;
  }

  public static class A {

    public String a;
  }

  public static void main(String[] args) throws IOException {
    Fury fury = Fury.builder().withLanguage(Language.XLANG)
            .withRefTracking(true).build();
    fury.register(Test.class, "example.bar");
    fury.register(A.class, "example.foo");

    A a = new A();
    a.a = "123";
    Test test = new Test();
    test.c = new A[]{a};
    byte[] bin = fury.serialize(test);

    Test deserialize = (Test)fury.deserialize(bin);
    System.out.println(deserialize.c);
  }
}

```
```Text
Exception in thread "main" io.fury.exception.InsecureException: class [Lio.fury.FuryTest$A; is not registered, if it's not the type you want to serialize, it may be a **vulnerability**. If it's not a vulnerability, registering class by `Fury#register` will have better performance, otherwise class name will be serialized too.
	at io.fury.resolver.ClassResolver.createSerializer(ClassResolver.java:1009)
	at io.fury.resolver.ClassResolver.getOrUpdateClassInfo(ClassResolver.java:961)
	at io.fury.resolver.ClassResolver.getSerializer(ClassResolver.java:677)
	at io.fury.type.GenericType.getSerializer(GenericType.java:184)
	at io.fury.type.GenericType.getSerializerOrNull(GenericType.java:196)
	at io.fury.serializer.StructSerializer.xwrite(StructSerializer.java:130)
	at io.fury.Fury.xwriteNonRef(Fury.java:426)
	at io.fury.Fury.xwriteRef(Fury.java:382)
	at io.fury.Fury.xserializeInternal(Fury.java:266)
	at io.fury.Fury.serialize(Fury.java:236)
	at io.fury.Fury.serialize(Fury.java:181)
	at io.fury.FuryTest.main(FuryTest.java:38)
```

@chaokunyang  I found an error while fixing this issue. Can you help take a look?

**Comment by wangweipeng2** [comment]
@3dot141  We have released version 0.0.16 of @furyjs/fury. Could you help us try it again?

**Comment by wangweipeng2** [comment]
> 
I will close the issue, feel free to open new issue if you encounter any issues
